### PR TITLE
PieWidgetUI: Refactor & New CentralText component

### PR DESCRIPTION
**Pull request by vmilan**
# Description

Shortcut: https://app.shortcut.com/cartoteam/story/346963/behaviour-of-central-text
[sc-346963] [sc-351761]

Echarts doesn't provide a proper way to extend the functionality of `Labels` or `Title`, so we need to create a new component to replace the text in the center of the chart to accomplish this product requirement: 

- The central text represents the most frequent category of the selection: whether they are all or just specific categories selected. Don't take into account deactivated categories.
- Don't update the central text and keep it as the most frequent category - Don't change the category on hover
- Add color indicator next to the category name

https://github.com/CartoDB/carto-react/assets/1934144/3cd56b22-faed-48af-877a-84c4ebeca83b

### CODE REVIEW

File: packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js [hunk 1]
```diff
@@ -98,38 +80,52 @@ function PieWidgetUI({
     [formatter, theme.palette.common.white, theme.palette.black, tooltipFormatter]
   );
 
-  // Series
```
Comment (vmilan) [inline packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js:101]: No longer needed, we use `CentralText` from now on

File: packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { useMemo, useRef, useEffect } from 'react';
+import { useTheme } from '@mui/material';
+import { OTHERS_CATEGORY_NAME, ORDER_TYPES } from '../../utils/chartConstants';
+import { sortDataDescending } from '../../utils/chartUtils';
+
+function usePieCategories(data, order, maxItems, colors) {
+  const theme = useTheme();
+  const colorByCategory = useRef({});
```
Comment (zbigg) [inline packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js:8]: I know it's old, but just raising awareness.

I wonder if keeping color mapping in ref makes sense. I would say, not.

This makes color mapping stable across data changes (apart from unmounting/remounting that resets ref, but that's other story)

I believe, If we change data we should always apply first color to first (or most dominant) category in _current_ data, not try to keep colors stable.

This is at least how i did it in TS widget.

What do you think?

So

File: packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }, [maxItems, orderedData]);
+
+  // Add a color to each category
+  const dataWithColor = useMemo(() => {
+    return groupedData.map(processDataItem(colorByCategory, colors, theme));
+  }, [groupedData, colors, theme]);
+
+  // Reset colorByCategory when colors and categories change
+  useEffect(() => {
+    colorByCategory.current = {};
```
Comment (zbigg) [inline packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js:54]: and this one is actually resetting `colorByCategory` on each change of data, so using ref is actually no-op

`colorByCategory` should be local to `dataWithColor` calculation

**Review by zbigg — APPROVED** [review]
LGTM, apart from weird and extra old logic which i believe can be fixed and some old code can be removed

File: packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js (hunk 2 above)
Comment (vmilan) [reply packages/react-ui/src/widgets/PieWidgetUI/hooks/usePieCategories.js]: Good catch, fixed here https://github.com/CartoDB/carto-react/pull/779/commits/505f6251f853852372ab87c5f83009bb04a8e0ad

File: packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js [hunk 4]
```diff
@@ -98,38 +80,52 @@ function PieWidgetUI({
...
+    let sumValue = 0;
+    for (const category of processedData) {
+      sumValue += category.value;
+    }
+
+    const percentage = calculatePercentage(category.value, sumValue);
+    category.percentage = percentage;
+
+    return category;
+  }, [processedData, selectedCategories]);
```
Comment (VictorVelarde) [inline packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js:114]: Didn't you consider putting this hook outside, as a useTopSelectedCategory?

**Review by VictorVelarde — APPROVED** [review]
LGTM

File: packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js (hunk 4 above)
Comment (vmilan) [reply packages/react-ui/src/widgets/PieWidgetUI/PieWidgetUI.js]: I moved to the `CentralText` component

**Pull request merged**